### PR TITLE
#184651064 user profile settings

### PR DESCRIPTION
**Pull request by Kress2000**
#### What does this PR do?
This is route ``users/settings`` for the user who has logged in, to edit, add, (Update) their information.
- Here ensures that the user is able to update his info but not others
- Ensures also that user has login first before accessing his info 
#### How should this be manually tested?
Manually, you can use swagger doc or postman to test, 
- here you go there and log in first
- then after login, you will have access to your settings where at /users/settings path, you will be able to add more info about yourself which are these: 
```
    username (not editable here),
    email  (not editable here),
    password  (not editable here),
    placeOfLiving,
    DOB,
    lang,
    currency,
    gender,
   tel,
  acc_number
```
According to the schema.
#### Any background context you want to provide?
Of course as the app, this is a PostgreSQL server, and it checks if the user is logged in first, using sessions.

#### What are the relevant pivotal tracker/Trello stories?
#184651064

### CODE REVIEW

File: src/controllers/user.controller.js [hunk 1]
```diff
@@ -16,11 +16,15 @@ const SignUp = async (req, res, next) => {
...
         return res.status(500).json({ error: err.message });
       }
-
       const body = {
         id: req.user.id,
         username: req.user.username,
         email: req.user.email,
+        gender: '',
+        currency: 'USD',
+        lang: 'EN',
```
Comment (didasmbalanya) [inline src/controllers/user.controller.js:25]: You can move these default values to the database models and migrations

File: src/controllers/userProfileSettings.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import User from '../models/user.model.js';
+import redisClient from '../helpers/redis.js';
+
+const { hash } = bcrypt;
+
+const UpdateUserInfo = async (req, res) => {
+  const userId = req.user.id;
+  const {
+    username,
+    // email,
```
Comment (didasmbalanya) [inline src/controllers/userProfileSettings.js:12]: Why is email commented out ?

File: src/controllers/userProfileSettings.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    password,
```
Comment (didasmbalanya) [inline src/controllers/userProfileSettings.js:13]: A user should not be able to update password like this. It needs the email flow and for our cases when they get notified that the password needs to be changed or when they need to reset

File: src/models/validationSchema.js [hunk 4]
```diff
@@ -17,12 +17,19 @@ const SignUpSchema = Joi.object().keys({
     .required()
     .messages(errorMessage('Username')),
   email: Joi.string().email().required().messages(errorMessage('Email')),
+  dob: Joi.date().messages(errorMessage('Date')),
+  gender: Joi.string().max(50).messages(errorMessage('Gender')),
+  currency: Joi.string().max(50).messages(errorMessage('Currency')),
+  placeofliving: Joi.string()
+    .max(225)
+    .messages(errorMessage('Place of living')),
+  lang: Joi.string().max(225).messages(errorMessage('Language')),
```
Comment (didasmbalanya) [inline src/models/validationSchema.js:26]: remove the validation schema file from the models folder and find a better place to put it. I know you are not the one who put it here but you can fix it

File: src/middleware/authentication.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import Jwt from 'jsonwebtoken';
-import redisClient from '../helpers/redis';
+import redisClient from '../helpers/redis.js';
```
Comment (didasmbalanya) [inline src/middleware/authentication.js:2]: Why add the extension ?

File: src/models/UserProfile.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+//   },
+// });
+// User.hasOne(UserProfile, {
+//   foreignKey: 'userId',
+//   onDelete: 'CASCADE',
+//   onUpdate: 'CASCADE',
+// });
+// UserProfile.belongsTo(User, { foreignKey: 'userId' });
+
+// export default UserProfile;
```
Comment (didasmbalanya) [inline src/models/UserProfile.js:51]: Do you expect this to be merged? If you prefer to have User profile as a separate model then discuss with your team and do so

File: src/models/user.model.js [hunk 7]
```diff
@@ -22,6 +22,26 @@ const User = sequelize.define('users', {
...
+    allowNull: true,
+  },
+  gender: {
+    type: Sequelize.STRING,
+    allowNull: true,
+  },
+  placeofliving: {
+    type: Sequelize.STRING,
+    allowNull: true,
+  },
```
Comment (didasmbalanya) [inline src/models/user.model.js:44]: Check the story. You might need more info for payment details for a user profile

File: src/tests/unit/userProfileSettings.test.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+const { expect } = chai;
+
+const { hash } = bcrypt;
+
+describe('PUT /settings', () => {
+  let token = '';
+  let userId = '';
+  let idx;
+  before(async () => {
+    sequelize.sync();
```
Comment (didasmbalanya) [inline src/tests/unit/userProfileSettings.test.js:23]: This is wrong. We won't be syncing the db from models. I need file migrations to create the table before all tests

**Review by didasmbalanya — CHANGES_REQUESTED** [review]

**Review by ishimwe5555 — CHANGES_REQUESTED** [review]

File: src/models/UserProfile.js (hunk 6 above)
Comment (Kress2000) [reply src/models/UserProfile.js:51]: i made it, in its separete schema called users, which will discuss with my team to use it as they will add some items a user can buy. and about model, we will use the same table as as it says user's info

File: src/controllers/user.controller.js (hunk 1 above)
Comment (Kress2000) [reply src/controllers/user.controller.js:25]: okay

File: src/controllers/userProfileSettings.js (hunk 2 above)
Comment (Kress2000) [reply src/controllers/userProfileSettings.js:12]: here email is not to be changed

File: src/controllers/userProfileSettings.js (hunk 3 above)
Comment (Kress2000) [reply src/controllers/userProfileSettings.js:13]: yes, let me do it

File: src/models/validationSchema.js (hunk 4 above)
Comment (Kress2000) [reply src/models/validationSchema.js:26]: yes let me do it!

File: src/middleware/authentication.js (hunk 5 above)
Comment (Kress2000) [reply src/middleware/authentication.js:2]: i was missig the imported server, in ES

File: src/models/user.model.js (hunk 7 above)
Comment (Kress2000) [reply src/models/user.model.js:44]: - About renaming ``placeofliving`` , is not needed, no problem with it, unless you suggest a better name;
- About payment details, I will created it in different table that references users table via users(id), but here is an issue @didasmbalanya ,  this is my task. if you agree with me that I am to do it this this suggestion of creating different table on criteria 2., I will need some more details such as payment details parameters(prices, addedItems, totalPrice, etc). This means that criteria number two would not go with this PR, I would doit later after @Edcreation id done with ``adding items to cart``, and me: @Kress2000  also finished ``removing items to cart`` and ``viewing added items``
image

Comment (Kress2000) [reply src/models/user.model.js:44]: but if it is only billing adress, means, it only to add new few columns such as ( account number, telnumber, etc, )

File: src/controllers/user.controller.js (hunk 1 above)
Comment (Kress2000) [reply src/controllers/user.controller.js:25]: done

File: src/controllers/userProfileSettings.js (hunk 2 above)
Comment (Kress2000) [reply src/controllers/userProfileSettings.js:12]: simply, i am not even allowed to change user's email, password, and username

File: src/controllers/userProfileSettings.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import redisClient from '../helpers/redis.js';
+
+/* eslint-disable camelcase */
+const UpdateUserInfo = async (req, res) => {
+  const userId = req.user.id;
+  const {
+    gender,
+    email,
+    username,
+    password,
```
Comment (ishimwe5555) [inline src/controllers/userProfileSettings.js:11]: Why did you select email, username and password? I think you don't need any if them in this case.

**Review by ishimwe5555 — CHANGES_REQUESTED** [review]

File: src/tests/unit/login.test.js [hunk 10]
```diff
@@ -135,24 +135,24 @@ describe('Testing User Login Routes with Errors', function () {
       .end((err, res) => {
         res.should.have.status(406);
         res.body.should.be.a('object');
-        res.body.should.have.property('message').eql('Password is incorrect');
+        res.body.should.have.property('message').eql('User not Found.');
```
Comment (ishimwe5555) [inline src/tests/unit/login.test.js:138]: Why did you edit this test? I think it's supposed to test trying to log in with an incorrect password.

**Review by ishimwe5555 — CHANGES_REQUESTED** [review]

File: src/controllers/userProfileSettings.js (hunk 9 above)
Comment (Kress2000) [reply src/controllers/userProfileSettings.js:11]: No! just check well down I am using them though, I am not editing them

File: .eslintignore [hunk 11]
```diff
@@ -7,3 +7,4 @@ package-lock.json
 .husky
 .github
 .github/*
+src/database/models/index.js
```
[the remark this answers is not in the document]
Comment (Kress2000) [reply .eslintignore:10]: my changes and @ishimwe5555's changes has merged the way comlicated! but now, i dit git reset to return on my own last commit, and these changes will be solved in PR

File: src/controllers/userProfileSettings.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
+import User from '../models/user.model.js';
+import redisClient from '../helpers/redis.js';
+
+/* eslint-disable camelcase */
```
[the remark this answers is not in the document]
Comment (Kress2000) [reply src/controllers/userProfileSettings.js:4]: this was done by me actually, and I did this to  avoid this poping error on the original name from DB,  but I had to put it on that line not the whole file, let me do it.

File: src/helpers/authHelper.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
+import User from '../models/user.model';
+
+// eslint-disable-next-line no-unused-vars
```
[the remark this answers is not in the document]
Comment (Kress2000) [reply src/helpers/authHelper.js:3]: this is to be changed too

File: src/controllers/userProfileSettings.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    password,
+    currency,
+    lang,
+    dob,
+    placeofliving,
+    tel,
+    acc_number,
+  } = req.body;
+  try {
+    const user = await User.findOne({ where: { id: userId } });
```
[the remark this answers is not in the document]
Comment (Kress2000) [reply src/controllers/userProfileSettings.js:20]: So,  does it mean, we will create a middleware that checks if a user is registered?

**Pull request closed without merging**